# **Project 4: LLM Project Activity - Topic Modeling**
### **Week 23** 2-Representation
Apply tokenization and text representation methods in the project. (9.3)

- First, convert tokens from dataset back to string for feature extraction
- Chosen TF-IDF for applying feature extration
- One complete, will then build a model using Logistic Regression
- Finally, will test the performance of the model against test data

In [ ]:
#Convert tokenized text back to string
def detokenize(text):
    if isinstance(text, list):
        return ' '.join(text)
    return text

new_ds = new_ds.map(lambda example: {'text': detokenize(example['text'])})

train_texts = new_ds['train']['text']
test_texts = new_ds['test']['text']

Map:   0%|          | 0/11314 [00:00<?, ? examples/s]

Map:   0%|          | 0/7532 [00:00<?, ? examples/s]

In [ ]:
#TF-IDF application for feature extraction
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()
X_train = vectorizer.fit_transform(train_texts)
X_test = vectorizer.transform(test_texts)


In [ ]:
#Prepare lablels
y_train = new_ds['train']['label']
y_test = new_ds['test']['label']

In [ ]:
#Train Logistic Regression Model
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.49      0.45      0.47       319
           1       0.60      0.70      0.65       389
           2       0.66      0.62      0.64       394
           3       0.66      0.62      0.64       392
           4       0.77      0.68      0.72       385
           5       0.80      0.70      0.75       395
           6       0.72      0.79      0.76       390
           7       0.77      0.70      0.73       396
           8       0.48      0.81      0.60       398
           9       0.81      0.79      0.80       397
          10       0.88      0.87      0.87       399
          11       0.86      0.67      0.75       396
          12       0.54      0.59      0.56       393
          13       0.76      0.76      0.76       396
          14       0.71      0.74      0.72       394
          15       0.64      0.79      0.70       398
          16       0.58      0.69      0.63       364
          17       0.85    

**Model Performance Summary**

Overall Accuracy: 68%

F1-scores: performance varies significantly across classes (e.g., classes 10,6 demonstrate strong performance ~.87-.76, other classes 18,0 demonstrate weaker performance ~.49-.47)

Results suggest possible class imbalance (e.g., class 19 had 251 samples/scored lowest F1=.26 vs. class 2 had 394 samples/scored better F1=.64)

Also, potential for false positives as some classes (e.g., 8) showed high recall but low precision.


